src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -0,0 +1,195 @@
+     * @return useCache
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    private boolean decideAndUseCache(String remoteName) throws IOException, InterruptedException {
+        boolean useCache = false;
+        String cacheEntry = getCacheEntry(remoteName);
+        File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry);
+        if (cacheDir != null) {
+            Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
+            GitClient client = git.getClient();
+            if (client.hasGitRepo()) {
+                sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
+                sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
+                useCache = true;
+            }
+        }
+        return useCache;
+    }
+
+    private void decideAndUseAPI(String remoteName, Item context, String credentialsId, String gitExe) {
+        if (setSizeFromAPI(remoteName, context, credentialsId)) {
+            implementation = determineSwitchOnSize(sizeOfRepo, gitExe);
+        }
+    }
+
+    /**
+     * Check if the desired implementation of extension is present and ask for the size of repository if it does
+     * @param repoUrl: The remote name derived from {@link GitSCMSource} object
+     * @return boolean useAPI or not.
+     */
+    private boolean setSizeFromAPI(String repoUrl, Item context, String credentialsId) {
+        List<RepositorySizeAPI> acceptedRepository = Objects.requireNonNull(RepositorySizeAPI.all())
+                .stream()
+                .filter(r -> r.isApplicableTo(repoUrl, context, credentialsId))
+                .collect(Collectors.toList());
+
+        if (acceptedRepository.size() > 0) {
+            try {
+                for (RepositorySizeAPI repo: acceptedRepository) {
+                    long size = repo.getSizeOfRepository(repoUrl, context, credentialsId);
+                    if (size != 0) { sizeOfRepo = size; }
+                }
+            } catch (Exception e) {
+                LOGGER.log(Level.INFO, "Not using performance improvement from REST API: " + e.getMessage());
+                return false;
+            }
+            return sizeOfRepo != 0; // Check if the size of the repository is zero
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Recommend a git implementation on the basis of the given size of a repository
+     * @param sizeOfRepo: Size of a repository (in KiBs)
+     * @return a git implementation, "git" or "jgit"
+     */
+    private String determineSwitchOnSize(Long sizeOfRepo, String gitExe) {
+        if (sizeOfRepo != 0L) {
+            if (sizeOfRepo >= SIZE_TO_SWITCH) {
+                return determineToolName(gitExe, "git");
+            } else {
+                return determineToolName(gitExe, JGitTool.MAGIC_EXENAME);
+            }
+        }
+        return "NONE";
```

Comment:
If you were to switch to returning a `GitTool` you could use the NullObject pattern where you have a "null" `GitTool` that's set up for this class.

Reply:
That is a valid suggestion, however I am not sure if adding the GitTool class will over-complicate the current design or make it neater. 

Returning strings instead of a concrete class seems okay till we return `NONE`. Returning NONE means that we will have to add a check before feeding the value to git client initialiser.

Reply:
Don't you have to check regardless?

Reply:
Well functionally, we are not affected by it. I maybe wrong but I don't seem to recall if we have discussed the possibility of using a GitTool object instead of a String object as the returned object.